docs/security.html

```diff
@@ -257,93 +322,191 @@
+    sasl.enabled.mechanisms=PLAIN</pre></li>
+        </ol>
+      </li>
+
+      <li><h5><a id="security_sasl_plain_clientconfig" href="#security_sasl_plain_clientconfig">Configuring Kafka Clients</a></h5>
+        To configure SASL authentication  on the clients:
+        <ol>
+          <li>The <tt>KafkaClient</tt> section describes how the clients like producer and consumer can connect to the Kafka Broker. The following is an example configuration for a client for the PLAIN mechanism:
+            <pre>
+    KafkaClient {
+        org.apache.kafka.common.security.plain.PlainLoginModule required
+        username="alice"
+        password="alice-secret";
+    };</pre>
+          </li>
+          <li>Pass the JAAS config file location as JVM parameter to each client JVM:
+            <pre>    -Djava.security.auth.login.config=/etc/kafka/kafka_client_jaas.conf</pre></li>
+          <li>Configure the following properties in producer.properties or consumer.properties:
+            <pre>    security.protocol=SASL_SSL
+    sasl.mechanism=PLAIN</pre></li>
+        </ol>
+      </li>
+      <li><h5><a id="security_sasl_plain_production" href="#security_sasl_plain_production">Use of SASL/PLAIN in production</a></h5>
         <ul>
-            <li>Incrementally bounce the cluster nodes to open additional secured port(s).</li>
-            <li>Restart clients using the secured rather than PLAINTEXT port (assuming you are securing the client-broker connection).</li>
-            <li>Incrementally bounce the cluster again to enable broker-to-broker security (if this is required)</li>
-            <li>A final incremental bounce to close the PLAINTEXT port.</li>
+          <li>SASL/PLAIN should be used only with SSL as transport layer to ensure that clear passwords are not transmitted on the wire without encryption.</li>
+          <li>The default implementation of SASL/PLAIN in Kafka specifies usernames and passwords in the JAAS configuration file as shown
+            <a href="#security_sasl_plain_brokerconfig">here</a>. To avoid storing passwords on disk, you can plugin your own implementation of
+            <code>javax.security.auth.spi.LoginModule</code> that provides usernames and passwords from an external source. The login module implementation should
+            provide username as the public credential and password as the private credential of the <code>Subject</code>. The default implementation
+            <code>org.apache.kafka.common.security.plain.PlainLoginModule</code> can be used as an example.</li>
+          <li>In production systems, external authentication servers may implement password authentication. Kafka brokers can be integrated with these servers by adding
+            your own implementation of <code>javax.security.sasl.SaslServer</code>. The default implementation included in Kafka in the package
+            <code>org.apache.kafka.common.security.plain</code> can be used as an example to get started.</li>
```

Comment:
Could we describe how the user can plug in a customized PLAIN implementation (e.g., configuring the provider)?